File: Scripts/GameStateManager.cs
using UnityEngine;

/// <summary>
/// GameStateManager - Central authority for all game state management
/// Tracks match time (90 minutes scaled to 6 real minutes), score, possession, and game flow states
/// 
/// TIME SCALING: 90 real minutes = 6 real-time minutes (15:1 ratio)
/// This keeps arcade gameplay fast while maintaining realistic match structure
/// </summary>
public class GameStateManager : MonoBehaviour
{
    // ===== Singleton Pattern =====
    public static GameStateManager Instance { get; private set; }

    // ===== Game State Enum =====
    public enum GameState
    {
        KickOff,      // Match start or after goal
        Gameplay,     // Normal play
        ThrowIn,      // Out of bounds on sideline
        Penalty,      // Penalty kick
        CornerKick,   // Corner kick
        GoalKick,     // Goal kick
        Paused,       // Game paused
        HalfTime,     // Between halves
        MatchEnd      // Match finished
    }

    // ===== Time Management =====
    private const float REAL_MATCH_DURATION = 360f;  // 6 real minutes in seconds
    private const float SIMULATED_MATCH_DURATION = 5400f; // 90 real match minutes in seconds
    private const float TIME_SCALE_RATIO = SIMULATED_MATCH_DURATION / REAL_MATCH_DURATION; // 15:1

    private float matchElapsedTime = 0f;        // Total elapsed time in game time (seconds)
    private bool isFirstHalf = true;
    private bool isMatchActive = false;
    private bool isPaused = false;

    // ===== Score Management =====
    [SerializeField] private int teamAScore = 0;
    [SerializeField] private int teamBScore = 0;

    // ===== Possession Tracking =====
    public enum Team { TeamA, TeamB, None }
    private Team currentPossession = Team.None;
    private float possessionChangeTime = 0f;

    // ===== Game State =====
    private GameState currentGameState = GameState.KickOff;
    private GameState previousGameState = GameState.KickOff;

    // ===== Events =====
    public delegate void OnGameStateChanged(GameState newState, GameState oldState);
    public event OnGameStateChanged GameStateChangedEvent;

    public delegate void OnScoreChanged(int teamAScore, int teamBScore);
    public event OnScoreChanged ScoreChangedEvent;

    public delegate void OnPossessionChanged(Team newTeam, Team oldTeam);
    public event OnPossessionChanged PossessionChangedEvent;

    public delegate void OnMatchTimeChanged(float matchTime);
    public event OnMatchTimeChanged MatchTimeChangedEvent;

    // ===== Initialization =====
    private void Awake()
    {
        // Enforce Singleton pattern
        if (Instance != null && Instance != this)
        {
            Destroy(gameObject);
            return;
        }

        Instance = this;
        DontDestroyOnLoad(gameObject);
    }

    // ===== Update Loop =====
    private void Update()
    {
        if (!isMatchActive || isPaused)
            return;

        // Increment match time scaled to real-time (15x speed)
        matchElapsedTime += Time.deltaTime * TIME_SCALE_RATIO;

        // Check for half-time or match end
        float halfDuration = SIMULATED_MATCH_DURATION / 2f;
        if (isFirstHalf && matchElapsedTime >= halfDuration)
        {
            HandleHalfTime();
        }
        else if (!isFirstHalf && matchElapsedTime >= SIMULATED_MATCH_DURATION)
        {
            HandleMatchEnd();
        }

        // Broadcast time change event
        MatchTimeChangedEvent?.Invoke(matchElapsedTime);
    }

    // ===== Match Control =====
    public void StartMatch()
    {
        isMatchActive = true;
        isPaused = false;
        matchElapsedTime = 0f;
        isFirstHalf = true;
        SetGameState(GameState.KickOff);
        Debug.Log("Match Started!");
    }

    public void PauseMatch()
    {
        isPaused = true;
        SetGameState(GameState.Paused);
        Debug.Log("Match Paused");
    }

    public void ResumeMatch()
    {
        if (currentGameState == GameState.Paused)
        {
            isPaused = false;
            SetGameState(previousGameState);
            Debug.Log("Match Resumed");
        }
    }

    public void EndMatch()
    {
        isMatchActive = false;
        SetGameState(GameState.MatchEnd);
        Debug.Log($"Match Ended! Final Score - Team A: {teamAScore} vs Team B: {teamBScore}");
    }

    // ===== Game State Management =====
    public void SetGameState(GameState newState)
    {
        if (currentGameState == newState)
            return;

        previousGameState = currentGameState;
        currentGameState = newState;

        Debug.Log($"Game State Changed: {previousGameState} -> {currentGameState}");
        GameStateChangedEvent?.Invoke(currentGameState, previousGameState);
    }

    private void HandleHalfTime()
    {
        isFirstHalf = false;
        SetGameState(GameState.HalfTime);
        Debug.Log("Half Time! First half ended.");
        // Resume with kick-off after delay
        Invoke(nameof(ResumeFromHalfTime), 5f);
    }

    private void ResumeFromHalfTime()
    {
        matchElapsedTime = SIMULATED_MATCH_DURATION / 2f; // Start second half timer
        SetGameState(GameState.KickOff);
        Debug.Log("Second Half Starting!");
    }

    private void HandleMatchEnd()
    {
        isMatchActive = false;
        SetGameState(GameState.MatchEnd);
        Debug.Log("Match Complete!");
    }

    // ===== Score Management =====
    public void AddGoal(Team scoringTeam)
    {
        if (scoringTeam == Team.TeamA)
            teamAScore++;
        else if (scoringTeam == Team.TeamB)
            teamBScore++;

        Debug.Log($"GOAL! {scoringTeam} scores. Current Score - Team A: {teamAScore} vs Team B: {teamBScore}");
        ScoreChangedEvent?.Invoke(teamAScore, teamBScore);

        // Reset to kick-off
        SetGameState(GameState.KickOff);
    }

    // ===== Possession Management =====
    public void ChangePossession(Team newTeam)
    {
        if (currentPossession == newTeam)
            return;

        Team oldTeam = currentPossession;
        currentPossession = newTeam;
        possessionChangeTime = matchElapsedTime;

        Debug.Log($"Possession changed to: {currentPossession}");
        PossessionChangedEvent?.Invoke(currentPossession, oldTeam);
    }

    // ===== Getters =====
    public GameState GetCurrentGameState() => currentGameState;
    public Team GetCurrentPossession() => currentPossession;
    public float GetMatchTimeInSeconds() => matchElapsedTime;
    
    /// <summary>
    /// Returns match time in MM:SS format (simulated match minutes)
    /// </summary>
    public string GetMatchTimeFormatted()
    {
        int minutes = (int)(matchElapsedTime / 60f);
        int seconds = (int)(matchElapsedTime % 60f);
        return $"{minutes:D2}:{seconds:D2}";
    }

    public int GetTeamAScore() => teamAScore;
    public int GetTeamBScore() => teamBScore;
    public bool IsFirstHalf() => isFirstHalf;
    public bool IsMatchActive() => isMatchActive;
    public bool IsMatchPaused() => isPaused;
    public float GetPossessionDuration() => matchElapsedTime - possessionChangeTime;
}

File: GameStateManager.js
/**
 * GameStateManager.js - Central authority for all game state management
 * Tracks match time (90 minutes scaled to 6 real minutes), score, possession, and game flow states
 * 
 * TIME SCALING: 90 real minutes = 6 real-time minutes (15:1 ratio)
 * This keeps arcade gameplay fast while maintaining realistic match structure
 */

class GameStateManager {
    constructor() {
        // ===== Singleton Pattern =====
        if (GameStateManager.instance) {
            return GameStateManager.instance;
        }
        GameStateManager.instance = this;

        // ===== Game State Enum =====
        this.GameState = {
            KickOff: 'KickOff',
            Gameplay: 'Gameplay',
            ThrowIn: 'ThrowIn',
            Penalty: 'Penalty',
            CornerKick: 'CornerKick',
            GoalKick: 'GoalKick',
            Paused: 'Paused',
            HalfTime: 'HalfTime',
            MatchEnd: 'MatchEnd'
        };

        // ===== Team Enum =====
        this.Team = {
            TeamA: 'TeamA',
            TeamB: 'TeamB',
            None: 'None'
        };

        // ===== Time Management =====
        this.REAL_MATCH_DURATION = 360;        // 6 real minutes in seconds
        this.SIMULATED_MATCH_DURATION = 5400;  // 90 simulated match minutes in seconds
        this.TIME_SCALE_RATIO = this.SIMULATED_MATCH_DURATION / this.REAL_MATCH_DURATION; // 15:1

        this.matchElapsedTime = 0;
        this.isFirstHalf = true;
        this.isMatchActive = false;
        this.isPaused = false;

        // ===== Score Management =====
        this.teamAScore = 0;
        this.teamBScore = 0;

        // ===== Possession Tracking =====
        this.currentPossession = this.Team.None;
        this.possessionChangeTime = 0;

        // ===== Game State =====
        this.currentGameState = this.GameState.KickOff;
        this.previousGameState = this.GameState.KickOff;

        // ===== Events =====
        this.listeners = {
            gameStateChanged: [],
            scoreChanged: [],
            possessionChanged: [],
            matchTimeChanged: [],
            halfTime: [],
            matchEnd: []
        };
    }

    // ===== Event System =====
    on(eventName, callback) {
        if (this.listeners[eventName]) {
            this.listeners[eventName].push(callback);
        }
    }

    emit(eventName, ...args) {
        if (this.listeners[eventName]) {
            this.listeners[eventName].forEach(callback => callback(...args));
        }
    }

    // ===== Update Loop =====
    update(deltaTime) {
        if (!this.isMatchActive || this.isPaused) return;

        // Increment match time scaled to real-time (15x speed)
        this.matchElapsedTime += deltaTime * this.TIME_SCALE_RATIO;

        // Check for half-time or match end
        const halfDuration = this.SIMULATED_MATCH_DURATION / 2;
        if (this.isFirstHalf && this.matchElapsedTime >= halfDuration) {
            this.handleHalfTime();
        } else if (!this.isFirstHalf && this.matchElapsedTime >= this.SIMULATED_MATCH_DURATION) {
            this.handleMatchEnd();
        }

        // Broadcast time change event
        this.emit('matchTimeChanged', this.matchElapsedTime);
    }

    // ===== Match Control =====
    startMatch() {
        this.isMatchActive = true;
        this.isPaused = false;
        this.matchElapsedTime = 0;
        this.isFirstHalf = true;
        this.setGameState(this.GameState.KickOff);
        console.log('Match Started!');
    }

    pauseMatch() {
        this.isPaused = true;
        this.setGameState(this.GameState.Paused);
        console.log('Match Paused');
    }

    resumeMatch() {
        if (this.currentGameState === this.GameState.Paused) {
            this.isPaused = false;
            this.setGameState(this.previousGameState);
            console.log('Match Resumed');
        }
    }

    endMatch() {
        this.isMatchActive = false;
        this.setGameState(this.GameState.MatchEnd);
        console.log(`Match Ended! Final Score - Team A: ${this.teamAScore} vs Team B: ${this.teamBScore}`);
        this.emit('matchEnd', this.teamAScore, this.teamBScore);
    }

    // ===== Game State Management =====
    setGameState(newState) {
        if (this.currentGameState === newState) return;

        this.previousGameState = this.currentGameState;
        this.currentGameState = newState;

        console.log(`Game State Changed: ${this.previousGameState} -> ${this.currentGameState}`);
        this.emit('gameStateChanged', this.currentGameState, this.previousGameState);
    }

    handleHalfTime() {
        this.isFirstHalf = false;
        this.setGameState(this.GameState.HalfTime);
        console.log('Half Time! First half ended.');
        this.emit('halfTime');
        
        // Resume with kick-off after delay
        setTimeout(() => this.resumeFromHalfTime(), 3000);
    }

    resumeFromHalfTime() {
        this.matchElapsedTime = this.SIMULATED_MATCH_DURATION / 2; // Start second half timer
        this.setGameState(this.GameState.KickOff);
        console.log('Second Half Starting!');
    }

    handleMatchEnd() {
        this.isMatchActive = false;
        this.setGameState(this.GameState.MatchEnd);
        console.log('Match Complete!');
        this.emit('matchEnd');
    }

    // ===== Score Management =====
    addGoal(scoringTeam) {
        if (scoringTeam === this.Team.TeamA) {
            this.teamAScore++;
        } else if (scoringTeam === this.Team.TeamB) {
            this.teamBScore++;
        }

        console.log(`GOAL! ${scoringTeam} scores. Current Score - Team A: ${this.teamAScore} vs Team B: ${this.teamBScore}`);
        this.emit('scoreChanged', this.teamAScore, this.teamBScore);

        // Reset to kick-off
        this.setGameState(this.GameState.KickOff);
    }

    // ===== Possession Management =====
    changePossession(newTeam) {
        if (this.currentPossession === newTeam) return;

        const oldTeam = this.currentPossession;
        this.currentPossession = newTeam;
        this.possessionChangeTime = this.matchElapsedTime;

        console.log(`Possession changed to: ${this.currentPossession}`);
        this.emit('possessionChanged', this.currentPossession, oldTeam);
    }

    // ===== Getters =====
    getCurrentGameState() {
        return this.currentGameState;
    }

    getCurrentPossession() {
        return this.currentPossession;
    }

    getMatchTimeInSeconds() {
        return this.matchElapsedTime;
    }

    /**
     * Returns match time in MM:SS format (simulated match minutes)
     */
    getMatchTimeFormatted() {
        const minutes = Math.floor(this.matchElapsedTime / 60);
        const seconds = Math.floor(this.matchElapsedTime % 60);
        return `${String(minutes).padStart(2, '0')}:${String(seconds).padStart(2, '0')}`;
    }

    getTeamAScore() {
        return this.teamAScore;
    }

    getTeamBScore() {
        return this.teamBScore;
    }

    isFirstHalfActive() {
        return this.isFirstHalf;
    }

    isMatchActiveState() {
        return this.isMatchActive;
    }

    isMatchPausedState() {
        return this.isPaused;
    }

    getPossessionDuration() {
        return this.matchElapsedTime - this.possessionChangeTime;
    }
}

// Export singleton instance
const gameStateManager = new GameStateManager();
